mesonbuild/optinterpreter.py

```diff
@@ -166,60 +198,90 @@ def func_option(self, args: T.Tuple[str], kwargs: 'FuncOptionArgs') -> None:
+            deprecated_values={'true': ('1.1.0', 'use a boolean, not a string'), 'false': ('1.1.0', 'use a boolean, not a string')},
+        ),
+        _YIELD_KW,
+    )
+    def boolean_parser(self, description: str, kwargs: BooleanArgs) -> coredata.UserOption:
+        return coredata.UserBooleanOption(description, kwargs['value'], kwargs['yield'])
 
-    @permittedKwargs({'value', 'min', 'max', 'yield'})
-    def integer_parser(self, description: str, kwargs: 'ParserArgs') -> coredata.UserOption:
-        value = kwargs.get('value')
+    @typed_kwargs(
+        'combo option',
+        KwargInfo('value', (str, NoneType)),
+        KwargInfo('choices', ContainerTypeInfo(list, str, allow_empty=False), required=True),
+        _YIELD_KW,
+    )
+    def combo_parser(self, description: str, kwargs: ComboArgs) -> coredata.UserOption:
+        choices = kwargs['choices']
+        value = kwargs['value']
         if value is None:
-            raise OptionException('Integer option must contain value argument.')
-        inttuple = (kwargs.get('min'), kwargs.get('max'), value)
+            value = kwargs['choices'][0]
+        return coredata.UserComboOption(description, choices, value, kwargs['yield'])
+
+    @typed_kwargs(
+        'integer option',
+        KwargInfo(
+            'value',
+            (int, str),
+            default=True,
+            validator=lambda x: None if isinstance(x, int) or x.isdigit() else 'integer values must be either an integer or a string representation thereof',
+            feature_validator=lambda x: [FeatureDeprecated('number values as strings', '1.1.0', 'use a raw number instead')] if isinstance(x, str) else [],
+            convertor=int,
+        ),
+        KwargInfo('min', (int, NoneType)),
+        KwargInfo('max', (int, NoneType)),
+        _YIELD_KW,
+    )
+    def integer_parser(self, description: str, kwargs: IntegerArgs) -> coredata.UserOption:
+        value = kwargs['value']
+        inttuple = (kwargs['min'], kwargs['max'], value)
         return coredata.UserIntegerOption(description, inttuple, kwargs['yield'])
 
-    @permittedKwargs({'value', 'yield', 'choices'})
-    def string_array_parser(self, description: str, kwargs: 'ParserArgs') -> coredata.UserOption:
-        choices = kwargs.get('choices', [])
-        value = kwargs.get('value', choices)
-        if not isinstance(value, list):
-            raise OptionException('Array choices must be passed as an array.')
+    @typed_kwargs(
+        'string array option',
+        KwargInfo('value', ContainerTypeInfo(list, str), default=[]),
+        KwargInfo('choices', ContainerTypeInfo(list, str), default=[]),
+        _YIELD_KW,
+    )
+    def string_array_parser(self, description: str, kwargs: StringArrayArgs) -> coredata.UserOption:
+        choices = kwargs['choices']
+        value = kwargs['value'] or choices
```

Comment:
I think default value should be None and use choices only in that case. An empty list could be the actual desired value.